### ISSUE TITLE
Issue with custom product attributes

### ISSUE THREAD

**Issue by SWS-Denzle** [woocommerce/woocommerce#15238]
### EXPLANATION OF THE ISSUE

Sometimes when adding & within a custom product attribute will show nothing (String gets removed from admin area and attribute shows blank on the front end). With the product I am playing with, adding Daryl & John will show nothing yet Daryl & Jpohn will. This store has been online for months now and has only turned out to be a recent issue.

To add, I have managed to recreate this on a testing system we have, if you would like to take a look, please let me know. The testing system only has WooCommerce enabled, no other plugins/themes are turned on.

### STEPS TO REPRODUCE THE ISSUE

Try and add Daryl & John within a custom product attribute.

### SYSTEM STATUS

Note: This is the testing systems status report not the live website (Which is also having this issue).
<details>
```
`
### WordPress Environment ###

Home URL: https://tst.maxvinylrecords.co.uk
Site URL: https://tst.maxvinylrecords.co.uk
WC Version: 3.0.7
Log Directory Writable: ✔
WP Version: 4.7.5
WP Multisite: –
WP Memory Limit: 2 GB
WP Debug Mode: –
WP Cron: ✔
Language: en_GB

### Server Environment ###

Server Info: Apache/2.4.25 (Unix) mod_hive/6.6 OpenSSL/1.0.1e-fips mod_fastcgi/2.4.6
PHP Version: 7.0.16
PHP Post Max Size: 32 MB
PHP Time Limit: 45
PHP Max Input Vars: 1000
cURL Version: 7.53.0
OpenSSL/1.0.2k

SUHOSIN Installed: –
MySQL Version: 5.6.36
Max Upload Size: 24 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.0.7
WC Database Prefix: mvwp_
woocommerce_sessions: ✔
woocommerce_api_keys: ✔
woocommerce_attribute_taxonomies: ✔
woocommerce_downloadable_product_permissions: ✔
woocommerce_order_items: ✔
woocommerce_order_itemmeta: ✔
woocommerce_tax_rates: ✔
woocommerce_tax_rate_locations: ✔
woocommerce_shipping_zones: ✔
woocommerce_shipping_zone_locations: ✔
woocommerce_shipping_zone_methods: ✔
woocommerce_payment_tokens: ✔
woocommerce_payment_tokenmeta: ✔
MaxMind GeoIP Database: ✔

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (2) ###

SG Optimizer: by SiteGround – 3.3.1
WooCommerce: by Automattic – 3.0.7

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: GBP (£)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Shop base: #3970 - /shop/
Basket: #9 - /basket/
Checkout: #10 - /checkout/
My account: #11 - /my-account/

### Theme ###

Name: Twenty Fifteen
Version: 1.7
Author URL: https://wordpress.org/
Child Theme: ❌ – If you're modifying WooCommerce on a parent theme you didn't build personally
then we recommend using a child theme. See: How to create a child theme

WooCommerce Support: ✔

### Templates ###

Overrides: –
`
```
</details>

**Comment by mikejolley** [comment]
No issue here https://dl.dropboxusercontent.com/s/dwkt6wpo9uhzpti/2017-05-23%20at%2020.21.png Email is on my profile if you want me to take a look.

**Comment by SWS-Denzle** [comment]
Hi @mikejolly,

If you don't mind that will be great, I've sent over the details to mike.jolley@me.com
I am at a loss as to what's happening to be honest.

Cheers

**Comment by SWS-Denzle** [comment]
Hi @mikejolley 

I've applied the fix you have mentioned above but we are still having problems at the moment and after more hours of searching, testing and looking, I am no closer to working this out.

It looks like for some reason when our customer edits some products it will remove all the attributes. When we look in the database as well, they are gone. 

If you try and add them again, it seems fine but why are they going missing?

To add to this, we also have a function which adds product columns to the product page to display some of this information, this is somehow still working.

Do you have any other fixes relating to Attributes that could be causing this issue?

If not, can you take another look please?

Something is wrong and I have spent days trying to get this working correctly again. 
Nothing in PHP Error log or WP_Debug.

If you need anything from me, please let me know.

**Comment by mikejolley** [comment]
Maybe you applied the fix incorrectly?

**Comment by SWS-Denzle** [comment]
Hi Mike,

I've checked this and followed exactly what you have added here: https://github.com/woocommerce/woocommerce/commit/7f8add52a76bd68536d8c4c1c94eea8482b293c5

Comment out:
//$term = get_term_by( 'name', $option, $this->get_name() );

Replace with:
$term = get_term_by( 'name', esc_attr( $option ), $this->get_name() );

**Comment by mikejolley** [comment]
After doing that you'll need to save on a new product? I tested after this fix and it worked with your example values.

**Comment by SWS-Denzle** [comment]
Hi Mike,

The product that lost all its attributes will now accept special characters for me, but seems to still be an issue for my customer. After I applied your fix (As I noticed the latest update is not being released until the end of the month) I asked my customer to do some testing for me, here is the response: 

**'We spoke yesterday about the Attributes.

Well I have successfully amended Rogers & Parton "Islands In The Stream" and it has accepted the & and all seems fine.

However, I have tried to amend "Come On Eileen" (MV8089)  and the attributes are still not showing for me to amend them.'**

When I try and edit "Come On Eileen" now myself, all the attributes are missing. If I then look in the database searching for this posts ID in the postmeta table, _product_attributes is empty ( showing  a:0:{} ).

If I now try and add an attribute at this point, it works fine.

So maybe the fix you supplied is working but something else is still not right somewhere. 
This is why I asked about any other fixes for attributes.

Based on the above, any ideas?

**Comment by mikejolley** [comment]
If that product was saved before the patch, the attributes would be missing.

**Comment by SWS-Denzle** [comment]
Hi Mike, 

Ok, I'll try and correct the ones at fault then to see if I have any issues with them and I'll ask our customer to test this again.

That should then hopefully nip this in the bud. 

Thank you and if its still an issue, I will let you know.

**Comment by SWS-Denzle** [comment]
Hi @mikejolley,

Just wanted to let you know that we found what the issue was. 
We still had attributes missing from products (Around 500) and the stange thing was, our custom function (Which runs in the admin area only) still found the data.

Looking in the database, it seems like the product attributes were being wrapped in a string within the database, i'll explain:

We found one of the attributes like this:

s:870:"a:5:{s:17:"pa_artist-surname";a:6:{s:4:"name";s:17:"pa_artist-surname";s:5:"value";s:0:"";s:8:"position";s:1:"1";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}s:12:"pa_condition";a:6:{s:4:"name";s:12:"pa_condition";s:5:"value";s:0:"";s:8:"position";s:1:"2";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}s:14:"pa_record-type";a:6:{s:4:"name";s:14:"pa_record-type";s:5:"value";s:0:"";s:8:"position";s:1:"3";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}s:14:"pa_record-year";a:6:{s:4:"name";s:14:"pa_record-year";s:5:"value";s:0:"";s:8:"position";s:1:"4";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}s:15:"pa_record-label";a:6:{s:4:"name";s:15:"pa_record-label";s:5:"value";s:0:"";s:8:"position";s:1:"5";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}}";
a:5:{s:17:"pa_artist-surname";a:6:{s:4:"name";s:17:"pa_artist-surname";s:5:"value";s:0:"";s:8:"position";s:1:"1";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}s:12:"pa_condition";a:6:{s:4:"name";s:12:"pa_condition";s:5:"value";s:0:"";s:8:"position";s:1:"2";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}s:14:"pa_record-type";a:6:{s:4:"name";s:14:"pa_record-type";s:5:"value";s:0:"";s:8:"position";s:1:"3";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}s:14:"pa_record-year";a:6:{s:4:"name";s:14:"pa_record-year";s:5:"value";s:0:"";s:8:"position";s:1:"4";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}s:15:"pa_record-label";a:6:{s:4:"name";s:15:"pa_record-label";s:5:"value";s:0:"";s:8:"position";s:1:"5";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}}

Comparing it with one that worked, we noticed  ' **s:870:"** ' and ' **";** '  wrapped around it. 
It should of been this:

a:5:{s:17:"pa_artist-surname";a:6:{s:4:"name";s:17:"pa_artist-surname";s:5:"value";s:0:"";s:8:"position";s:1:"1";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}s:12:"pa_condition";a:6:{s:4:"name";s:12:"pa_condition";s:5:"value";s:0:"";s:8:"position";s:1:"2";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}s:14:"pa_record-type";a:6:{s:4:"name";s:14:"pa_record-type";s:5:"value";s:0:"";s:8:"position";s:1:"3";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}s:14:"pa_record-year";a:6:{s:4:"name";s:14:"pa_record-year";s:5:"value";s:0:"";s:8:"position";s:1:"4";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}s:15:"pa_record-label";a:6:{s:4:"name";s:15:"pa_record-label";s:5:"value";s:0:"";s:8:"position";s:1:"5";s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;}}

We found several like this and after removing the combinations of  ' **s:870:"** ' and ' **";** '  (The 870 was differen't values) all the attributes came back.

Hopefully this is some use to someone if they should find them selves having this issue.

Cheers
